pkg/analysis/statussubresource/analyzer.go

```diff
@@ -0,0 +1,160 @@
+func newAnalyzer(cfg config.StatusSubresourceConfig) *analysis.Analyzer {
+	defaultConfig(&cfg)
+
+	a := &analyzer{
+		statusFieldName: cfg.StatusFieldName,
+	}
+
+	return &analysis.Analyzer{
+		Name:     name,
+		Doc:      "Checks that a type marked with kubebuilder:object:root:=true and containing a status field is marked with kubebuilder:subresource:status",
+		Run:      a.run,
+		Requires: []*analysis.Analyzer{inspect.Analyzer, markers.Analyzer},
+	}
+}
+
+func (a *analyzer) run(pass *analysis.Pass) (interface{}, error) {
+	inspect, ok := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
+	if !ok {
+		return nil, errCouldNotGetInspector
+	}
+
+	markersAccess, ok := pass.ResultOf[markers.Analyzer].(markers.Markers)
+	if !ok {
+		return nil, errCouldNotGetMarkers
+	}
+
+	// Filter to structs so that we can iterate over fields in a struct.
+	nodeFilter := []ast.Node{
+		(*ast.TypeSpec)(nil),
+	}
+
+	inspect.Preorder(nodeFilter, func(n ast.Node) {
+		typeSpec, ok := n.(*ast.TypeSpec)
+		if !ok {
+			return
+		}
+
+		sTyp, ok := typeSpec.Type.(*ast.StructType)
+		if !ok {
+			return
+		}
+
+		// no identifier on the type
+		if typeSpec.Name == nil {
+			return
+		}
+
+		structMarkers := markersAccess.StructMarkers(sTyp)
+		a.checkStruct(pass, sTyp, typeSpec.Name.Name, structMarkers)
+	})
+
+	return nil, nil //nolint:nilnil
+}
+
+func (a *analyzer) checkStruct(pass *analysis.Pass, sTyp *ast.StructType, name string, structMarkers markers.MarkerSet) {
+	if sTyp == nil {
+		return
+	}
+
+	if !structMarkers.Has(KubebuilderRootMarker) {
+		return
+	}
+
+	hasStatusSubresourceMarker := structMarkers.Has(KubebuilderStatusSubresourceMarker)
+	hasStatusField := hasStatusField(sTyp.Fields, a.statusFieldName)
+
+	switch {
+	case (hasStatusSubresourceMarker && hasStatusField), (!hasStatusSubresourceMarker && !hasStatusField):
+		// acceptable state
+	case hasStatusSubresourceMarker && !hasStatusField:
+		// Might be able to have some suggested fixes here, but it is likely much more complex
+		// so for now leave it with a descriptive failure message.
+		pass.Reportf(sTyp.Pos(), "root object type %q is marked to enable the status subresource with marker %q but has no status field", name, KubebuilderStatusSubresourceMarker)
+	case !hasStatusSubresourceMarker && hasStatusField:
+		// In this case we can suggest the autofix to add the status subresource marker
+		pass.Report(analysis.Diagnostic{
+			Pos:     sTyp.Pos(),
+			Message: fmt.Sprintf("root object type %q has a status field but does not have the marker %q to enable the status subresource", name, KubebuilderStatusSubresourceMarker),
+			SuggestedFixes: []analysis.SuggestedFix{
+				{
+					Message: "should add the kubebuilder:subresource:status marker",
+					TextEdits: []analysis.TextEdit{
+						// go one line above the struct and add the marker
+						{
+							Pos: sTyp.Pos() - token.Pos(len(name)) - token.Pos(2) - token.Pos(len("type")) - token.Pos(1),
```

Comment:
Raw inserts can be achieved by setting `Pos` to the `sTyp.Pos()`, and setting `End` to `token.NoPos`. I think this will achieve what you want here

Reply:
I'll play around with this a bit more and see if I can get something cleaner. The key thing here is that I want the insert to be on the line above the struct

Reply:
The conditions linter I'm working on presently has an example of this same thing, and it's working as I'm expecting, that's how I knew to comment here. LMK if you can't get it working and I'll send you a snippet

Reply:
I couldn't get what you mentioned to work for my case. A snippet would be helpful, but setting:
```go
{
    Pos: sTyp.Pos(),
    End: token.NoPos,
    ...
}
```
resulted in this:
```
// +kubebuilder:object:root:=true
type Baz 
// +kubebuilder:subresource:statusstruct {
```

I modified my original logic to be a bit more straightforward and explained it with a comment. With my existing logic I get the expected outcome of:
```
// +kubebuilder:object:root:=true
// +kubebuilder:subresource:status
type Baz struct {
```